packages/lib/restrictionSchedule.ts

```diff
@@ -0,0 +1,97 @@
+import dayjs from "@calcom/dayjs";
+import { FeaturesRepository } from "@calcom/features/flags/features.repository";
+
+export interface RestrictionScheduleAvailability {
+  days: number[];
+  startTime: Date;
+  endTime: Date;
+  date: Date | null;
+}
+
+export interface RestrictionSchedule {
+  id: number;
+  timeZone: string | null;
+  availability: RestrictionScheduleAvailability[];
+}
+
+export interface RestrictionScheduleCheckParams {
+  restrictionSchedule: RestrictionSchedule;
+  bookingStartTime: Dayjs;
+  bookingEndTime: Dayjs;
+  useBookerTimezone: boolean;
+  bookerTimezone?: string;
+}
+
+/**
+ * Checks if the restriction schedule feature is enabled for a team
+ * @param teamId - The team ID to check
+ * @returns Promise<boolean> - True if the feature is enabled, false otherwise
+ */
+export async function isRestrictionScheduleEnabled(teamId?: number): Promise<boolean> {
+  if (!teamId) {
+    return false; // Personal events don't have restriction schedules
+  }
+
+  const featureRepo = new FeaturesRepository();
+  return await featureRepo.checkIfTeamHasFeature(teamId, "restriction-schedule");
+}
+/**
+ * Checks if a booking time falls within the restriction schedule's availability
+ * @param params - The restriction schedule check parameters
+ * @returns true if the booking is allowed, false otherwise
+ * @throws Error if required timezone information is missing
+ */
+export function isBookingAllowedByRestrictionSchedule({
+  restrictionSchedule,
+  bookingStartTime,
+  bookingEndTime,
+  useBookerTimezone,
+  bookerTimezone,
+}: RestrictionScheduleCheckParams): boolean {
+  if (!restrictionSchedule.timeZone && !useBookerTimezone) {
+    throw new Error("No timezone is set for the restriction schedule and useBookerTimezone is false");
+  }
+
+  if (useBookerTimezone && !bookerTimezone) {
+    throw new Error("Booker timezone is required when useBookerTimezone is true");
+  }
+
+  const restrictionTimezone = useBookerTimezone ? bookerTimezone! : restrictionSchedule.timeZone!;
+
+  const dateOverrides = restrictionSchedule.availability.filter((a) => !!a.date);
+  const recurringRules = restrictionSchedule.availability.filter((a) => !a.date);
+
+  const bookingStartTimeNormalized = dayjs(bookingStartTime.utc()).tz(restrictionTimezone);
+  const bookingEndTimeNormalized = dayjs(bookingEndTime.utc()).tz(restrictionTimezone);
+  const bookingDateStr = bookingStartTimeNormalized.format("YYYY-MM-DD");
+  const bookingWeekday = bookingStartTimeNormalized.day();
+  const bookingStartValue = bookingStartTime.valueOf();
+  const bookingEndValue = bookingEndTime.valueOf();
+
+  const overrideRulesForTheDay = dateOverrides.filter((a) =>
+    dayjs.utc(a.date).isSame(bookingStartTime.utc(), "day")
+  );
+  const recurringRulesForTheDay = recurringRules.filter((a) => a.days.includes(bookingWeekday));
+
+  if (overrideRulesForTheDay.length > 0) {
+    const overrideAllowsBooking = overrideRulesForTheDay.some((rule) => {
```

Comment:
That's not working as expected, because one date override can merge with other date overrides like this: 
![Screenshot 2025-06-17 at 11 07 31 AM](https://github.com/user-attachments/assets/3826b4ab-1c27-4b5c-82ce-baf15445e17c)

This would not show a 30-minute slot, even though there is availability